src/NHibernate/Loader/Hql/QueryLoader.cs

```diff
@@ -495,6 +497,44 @@ internal IEnumerable GetEnumerable(QueryParameters queryParameters, IEventSource
 			return result;
 		}
 
+		internal IAsyncEnumerable<T> GetAsyncEnumerable<T>(QueryParameters queryParameters, IEventSource session)
+		{
+			CheckQuery(queryParameters);
+
+			return new AsyncEnumerableImpl<T>(
+				InitializeAsync,
+				queryParameters.IsReadOnly(session),
+				_queryTranslator.ReturnTypes,
+				_queryTranslator.GetColumnNames(),
+				queryParameters.RowSelection,
+				_selectNewTransformer ?? queryParameters.ResultTransformer,
+				_queryReturnAliases,
+				session);
+
+			async Task<InitializeResult> InitializeAsync(CancellationToken cancellationToken)
```

Comment:
With this the async enumerable becomes a "future"-like version. Which is not how the Enumerable behaves. I think we still need to execute query upfront, not on `MoveNextAsync()`

Reply:
>  I think we still need to execute query upfront, not on MoveNextAsync()

Initially I tried to copy `QueryLoader.GetEnumerable` behavior, but the problem is that in order to do that we would need to call `PrepareQueryCommand` and `GetResultSet` synchronously, otherwise we would have a similar issue as with the current `EnumerableAsync`, where the enumerable is wrapped in a task. As I am not sure what is the "correct" behavior, I've checked what Microsoft did with EF and they seem to execute the query on the first `MoveNextAsync`, calling `AsAsyncEnumerable` does not trigger any query (checked with SQL Serever Profiler). My current opinion on this is that `EnumerableImpl` is kind of broken as it does not support multiple iterations, which is also stated in the documentation:
https://github.com/nhibernate/nhibernate-core/blob/e8288a58c50a2ecf7ee1e47ee0e0f55c9a1213e5/src/NHibernate/Impl/EnumerableImpl.cs#L20-L23
and by executing the query upon calling `AsAsyncEnumerable`, means that we would only be able to iterate once.

Reply:
> My current opinion on this is that EnumerableImpl is kind of broken as it does not support multiple iterations, which is also stated in the documentation:

Agree. That's something I've been trying to fix in #2017.

Reply:
Unfurtunately #2017 does not fix multiple iterations as in order to support that the following code should be executed inside `EnumerableImpl` when first `MoveNext` is called, similar what was done for `AsyncEnumerableImpl` in this PR:
https://github.com/nhibernate/nhibernate-core/blob/d8931ac97e97f56d811ea8b5383c08e39a96df40/src/NHibernate/Loader/Hql/QueryLoader.cs#L478-L481

Reply:
> does not fix multiple iterations ... to support that the following code should be executed inside EnumerableImpl when first MoveNext is called,

Actually all I need to fix it is to comment these [2 lines](https://github.com/bahusoid/nhibernate-core/blob/0a796d5e85ce5a5d15bcf1119d6235c9782abbbd/src/NHibernate/Engine/Query/HQLQueryPlan.cs#L163-L164) in my PR. And case with multiple translators already works properly there.

Reply:
It got me thinking. Wouldn't it be less changes to implement async part similarly.
I mean keep `Translators.GetAsyncEnumerable` do query execution. And delay and make it reusable  using iterator
 similar to this [`GetEnumerable`](https://github.com/bahusoid/nhibernate-core/blob/0a796d5e85ce5a5d15bcf1119d6235c9782abbbd/src/NHibernate/Engine/Query/HQLQueryPlan.cs#L169-L178). Something like:
```C#
private IEnumerable<IAsyncEnumerable<object>> GetAsyncEnumerables(QueryParameters queryParameters, IEventSource session)
{
	foreach (var t in Translators)
	{
		foreach (object obj in t.GetAsyncEnumerable(queryParameters, session))
		{
			yield return obj;
		}
	}
}
```

Reply:
> Wouldn't it be less changes to implement async part similarly

This is what I was trying to achieve initially, but the problem is that without C#8 we are not able to use `await foreach`, which means we cannot make the following method:
```C#
private async IAsyncEnumerable<T> GetAsyncEnumerables<T>(QueryParameters queryParameters, IEventSource session)
{
  foreach (var t in Translators)
  {
    await foreach (var element in t.GetAsyncEnumerable<T>(queryParameters, session))
    {
      yield return element;
    }
  }
}
```
Your code would not compile because `IAsyncEnumerable` does not implement `IEnumerable` which is needed by `foreach` statement.

Reply:
That's why I suggest returning `IEnumerable<IAsyncEnumerable<T>` and not `IAsyncEnumerable<T>`. And "flattening" part should be done inside `JoinedAsyncEnumerable`

Reply:
> That's why I suggest returning IEnumerable<IAsyncEnumerable<T> and not IAsyncEnumerable<T>.

Now I understod what you meant and indeed it simplified `HQLQueryPlan.PerformAsyncIterate`.

> Actually all I need to fix it is to comment these 2 lines in my PR

That correct, but calling `IQueryTranslator.GetEnumerable` will still trigger the query to be executed, where `IQueryTranslator.GetAsyncEnumerable` in this PR does not. In order to make those two method consistent I deprecated `EnumerableImpl`, added `IEnumerable<>` to `AsyncEnumerableImpl` and used it instead.

Reply:
It doesn't add any big overheads? Then OK. So please confirm it makes #2017 obsolete

Reply:
> It doesn't add any big overheads for sync case?

Not really, the reading logic remained the same.

> So please confirm it makes #2017 obsolete

It does, I obsoleted `JoinedEnumerable` in this PR and adopted `PerformIterate` from your PR, but I kept the `if (Translators.Length == 1)` condition in order to avoid nested enumeration.